### PR TITLE
Fix generation of ARM markdown help files.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Include API specific name property instead of the generic one (it provides
better doc text than the generic "name of the resource" one).
Merge AllOf/Properties into a flat list so that AllOf is included.
Removed EnsureUniqueFileName as it's not needed.
Fix unnecessary nesting output directories.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@daschult for the changes in ResourceSchemaParser.cs

**Comment by daschult** [comment]
Looks fine to me, although these are changes mostly in the markdown area, which I don't know a lot about. It should be fine though.

**Comment by jhendrixMSFT** [comment]
@marstr can you take a quick look at the code changes?

File: src/generator/AutoRest.AzureResourceSchema/ResourceMarkdownGenerator.cs [hunk 1]
```diff
@@ -436,5 +450,46 @@ private JToken JsonSchemaToJTokenImpl(string propName, JsonSchema jsonSchema, St
...
+        /// <param name="items">The items to merge into the dictionary.</param>
+        private void FlattenPropertiesImpl(Dictionary<string, JsonSchema> merged, IEnumerable<JsonSchema> items)
+        {
+            foreach (var item in items)
+            {
+                if (item.AllOf != null)
+                {
+                    FlattenPropertiesImpl(merged, item.AllOf);
+                }
+                else if (item.Properties != null)
```
Comment (marstr) [inline src/generator/AutoRest.AzureResourceSchema/ResourceMarkdownGenerator.cs]: It's not clear to me that each item will have one of AllOf, Properties, or Ref to flatten. Are there any cases where multiple of these clauses would need to be hit?

File: src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs [hunk 2]
```diff
@@ -81,8 +81,28 @@ public static class ResourceSchemaParser
...
                     resourceDefinition.JsonType = "object";
-
                     resourceDefinition.ResourceType = resourceType;
+
+                    // get the resource name parameter, e.g. {fooName}
+                    var resNameParam = methodUrlPathAfterProvider.Substring(methodUrlPathAfterProvider.LastIndexOf('/') + 1);
+                    if (IsPathVariable(resNameParam))
+                    {
+                        // strip the enclosing braces
+                        resNameParam = resNameParam.Substring(1, resNameParam.Length - 2);
```
Comment (marstr) [inline src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs]: [nit] Perhaps a more readable version:

``` C#
resNameParam = resNameParam.Trim(new[]{'{', '}'});
```

File: src/generator/AutoRest.AzureResourceSchema/CodeGeneratorArs.cs [hunk 3]
```diff
@@ -29,35 +29,17 @@ public CodeGeneratorArs()
...
+                await Write(stringWriter.ToString(), resourceProvider + ".json", true);
 
                 stringWriter = new StringWriter();
                 var md = ResourceMarkdownGenerator.Generate(resourceSchemas[resourceProvider]);
 
                 foreach (var m in md)
                 {
                     var content = m.Content.Replace("\"boolean\"", "boolean");
-                    await Write(content, EnsureUniqueFileName(resourceProvider, m.Type + ".md"), false);
+                    await Write(content, m.Type + ".md", false);
```
Comment (jhendrixMSFT) [inline src/generator/AutoRest.AzureResourceSchema/CodeGeneratorArs.cs]: Yep m.Type is guaranteed to be unique.

File: src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs (hunk 2 above)
Comment (jhendrixMSFT) [inline src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs]: Good suggestion I'll make this change.

**Comment by jhendrixMSFT** [comment]
@fearthecowboy Ready for merging.

**Pull request merged**